src/OrleansRuntime/Core/InsideRuntimeClient.cs

```diff
@@ -414,6 +428,36 @@ public void SniffIncomingMessage(Message message)
             }
         }
 
+        private static void UnwrapCancellationToken(IAddressable target, object arg, InvokeMethodRequest request, int i)
+        {
+            var grainToken = ((GrainCancellationToken)arg);
+            if (grainToken.WentThroughSerialization)
+            {
+                CancellationSourcesExtension cancellationExtension;
+                if (!SiloProviderRuntime.Instance.TryGetExtensionHandler(out cancellationExtension))
+                {
+                    cancellationExtension = new CancellationSourcesExtension();
+
+                    if (!SiloProviderRuntime.Instance.TryAddExtension(cancellationExtension))
+                    {
+                        if (logger.IsWarning) logger.Warn(
+                            ErrorCode.CancellationExtensionCreationFailed,
+                            string.Format("Could not add cancellation token extension, target: {0}", target));
+                        return;
+                    }
+                }
+
+                request.Arguments[i] = cancellationExtension.GetOrCreateCancellationToken(
+                    grainToken.Id,
+                    grainToken.CancellationToken.IsCancellationRequested);
+            }
+            else
+            { 
+                // no need to register remote token extension for already cancelled or local token
```

Comment:
Non-local tokens which are cancelled will not arrive here, but the comment indicates that they will.